processing/src/main/java/io/druid/segment/StringDimensionHandler.java

```diff
@@ -0,0 +1,1609 @@
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Iterators;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.google.common.collect.Ordering;
+import com.google.common.io.ByteSink;
+import com.google.common.io.ByteStreams;
+import com.google.common.io.Files;
+import com.google.common.io.OutputSupplier;
+import com.google.common.primitives.Ints;
+import com.metamx.collections.bitmap.BitmapFactory;
+import com.metamx.collections.bitmap.ImmutableBitmap;
+import com.metamx.collections.bitmap.MutableBitmap;
+import com.metamx.collections.spatial.ImmutableRTree;
+import com.metamx.collections.spatial.RTree;
+import com.metamx.collections.spatial.split.LinearGutmanSplitStrategy;
+import com.metamx.common.ISE;
+import com.metamx.common.logger.Logger;
+import io.druid.collections.CombiningIterable;
+import io.druid.common.guava.FileOutputSupplier;
+import io.druid.common.utils.SerializerUtils;
+import io.druid.query.dimension.DimensionSpec;
+import io.druid.query.extraction.ExtractionFn;
+import io.druid.query.filter.DruidPredicateFactory;
+import io.druid.query.filter.ValueMatcher;
+import io.druid.segment.column.Column;
+import io.druid.segment.column.ColumnCapabilities;
+import io.druid.segment.column.ColumnDescriptor;
+import io.druid.segment.column.DictionaryEncodedColumn;
+import io.druid.segment.column.ValueType;
+import io.druid.segment.data.BitmapSerdeFactory;
+import io.druid.segment.data.ByteBufferWriter;
+import io.druid.segment.data.CompressedObjectStrategy;
+import io.druid.segment.data.CompressedVSizeIndexedV3Writer;
+import io.druid.segment.data.CompressedVSizeIntsIndexedWriter;
+import io.druid.segment.data.GenericIndexed;
+import io.druid.segment.data.GenericIndexedWriter;
+import io.druid.segment.data.IOPeon;
+import io.druid.segment.data.Indexed;
+import io.druid.segment.data.IndexedInts;
+import io.druid.segment.data.IndexedIntsWriter;
+import io.druid.segment.data.IndexedIterable;
+import io.druid.segment.data.IndexedRTree;
+import io.druid.segment.data.ListIndexed;
+import io.druid.segment.data.TmpFileIOPeon;
+import io.druid.segment.data.VSizeIndexedIntsWriter;
+import io.druid.segment.data.VSizeIndexedWriter;
+import io.druid.segment.filter.BooleanValueMatcher;
+import io.druid.segment.incremental.IncrementalIndex;
+import io.druid.segment.incremental.IncrementalIndexStorageAdapter;
+import io.druid.segment.serde.DictionaryEncodedColumnPartSerde;
+
+import javax.annotation.Nullable;
+import java.io.Closeable;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.nio.IntBuffer;
+import java.nio.MappedByteBuffer;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+public class StringDimensionHandler implements DimensionHandler<Integer, String>
```

Comment:
I think this file'd be easier to understand if broken up into at least four different files: handler, indexer, mergerv9, mergerlegacy. None of those share state with each other and they are doing pretty different jobs.